### ISSUE TITLE
Invalid runtime exception thrown when calling onStateChange for org.jsmpp.session.SMPPSession

### ISSUE THREAD

**Issue by cforno** [opentelecoms-org/jsmpp#89]
Hi, When creating a new SMPPSession I am receiving an exception. It seems that jsmpp is trying to initialize and threadpool executor with a values < 1, but I was not able to find in the source code of SMPP. After that exception I can continue usign session, but I am not sure if that can affect the normal operation of the session.

Code Used:
```
private SMPPSession newSession() throws IOException {
    	log.info("Connecting to "+remoteIpAddress+", port: "+remotePort+", transactionTimer: "+transactionTimer);
        SMPPSession tmpSession = new SMPPSession(remoteIpAddress, remotePort, bindParam);
        tmpSession.setTransactionTimer(transactionTimer);
        sessionStateListener = new SessionStateListenerImpl();
        messageReceiverListener = new MessageReceiverListenerImpl(connector);
        tmpSession.setMessageReceiverListener(messageReceiverListener);
        tmpSession.addSessionStateListener(sessionStateListener);
        tmpSession.setEnquireLinkTimer(120000);
        
        return tmpSession;
    }
```

```
private class SessionStateListenerImpl implements SessionStateListener {

		public void onStateChange(SessionState newState, SessionState oldState,
				Session session) {
			
			connector = connectorManager.get(connector.getId());
			if(connector.getStatus()){
				if (newState.equals(SessionState.CLOSED)) {
	            	log.error("Session to " + remoteIpAddress + ":" + remotePort+" closed, creating new session");
	                reconnectAfter(reconnectInterval);
	            }
			}else{
				session.removeSessionStateListener(sessionStateListener);
				session.unbindAndClose();
				session.close();
			}
		}
    }
```

Exception:
```
2018-08-24 14:23:36,088 [SMS-Conn-Id-40] ERROR (AbstractSessionContext.java:81) - Invalid runtime exception thrown when calling onStateChange for org.jsmpp.session.SMPPSession@2f20269
java.lang.IllegalArgumentException
	at java.base/java.util.concurrent.ThreadPoolExecutor.setCorePoolSize(ThreadPoolExecutor.java:1560)
	at org.jsmpp.session.SMPPSession$BoundSessionStateListener.onStateChange(SMPPSession.java:663)
	at org.jsmpp.session.AbstractSessionContext.fireStateChanged(AbstractSessionContext.java:79)
	at org.jsmpp.session.SMPPSessionContext.changeState(SMPPSessionContext.java:61)
	at org.jsmpp.session.AbstractSessionContext.bound(AbstractSessionContext.java:51)
	at org.jsmpp.session.SMPPSession.connectAndBind(SMPPSession.java:236)
	at org.jsmpp.session.SMPPSession.connectAndBind(SMPPSession.java:201)
	at org.jsmpp.session.SMPPSession.<init>(SMPPSession.java:140)
```

Version: Jsmpp 2.3.6
Thanks

**Comment by pmoerenhout** [comment]
The code which is failing is at SmppSession.java:662:
logger.info("Changing processor degree to {}", getPduProcessorDegree());               	((ThreadPoolExecutor)pduReaderWorker.executorService).setCorePoolSize(getPduProcessorDegree());
((ThreadPoolExecutor)pduReaderWorker.executorService).setMaximumPoolSize(getPduProcessorDegree());
The logging should show to what value the processor degree is set. The default is 3, but it should have a illegal value (0 ?) in your application. Could you share some more logging and/or code ?

**Comment by cforno** [comment]
Hi Pim, thank you for your response, please find attached jsmpp debug logs:

```
> 2018-08-27 10:10:54,895 [SMS-Conn-Id-83] INFO  (AutoReconnectGateway.java:116) - Connecting to 172.16.1.13, port: 7777, transactionTimer: 120000
> 2018-08-27 10:10:54,913 [SMS-Conn-Id-83] DEBUG (SMPPSession.java:217) - Connect and bind to 172.16.1.13 port 7777
> 2018-08-27 10:10:54,921 [SMS-Conn-Id-83] INFO  (SMPPSession.java:223) - Connected to /172.16.1.13
> 2018-08-27 10:10:54,922 [PDUReaderWorker: org.jsmpp.session.SMPPSession@491f0632] INFO  (SMPPSession.java:580) - Starting PDUReaderWorker
> 2018-08-27 10:10:54,925 [SMS-Conn-Id-83] DEBUG (DefaultPDUSender.java:450) - Sending SMPP message 00 00 00 2a 00 00 00 09 00 00 00 00 00 00 00 01 69 74 64 63 68 69 6e 6f 00 31 74 64 63 68 31 6e 6f 00 53 4d 53 00 34 01 01 00 
> 2018-08-27 10:10:54,931 [pool-7-thread-1] DEBUG (PDUProcessTask.java:54) - Received SMPP message PDUHeader(25, 80000009, 00000000, 1) 69 74 64 63 68 69 6e 6f 00 
> 2018-08-27 10:10:54,931 [pool-7-thread-1] DEBUG (SMPPSessionOpen.java:57) - Bind Response header (25, 80000009, 00000000, 1)
> 2018-08-27 10:10:54,932 [SMS-Conn-Id-83] DEBUG (AbstractSession.java:292) - bind response with sequence 1 received for session c1d3ad2f
> 2018-08-27 10:10:54,942 [SMS-Conn-Id-83] INFO  (OptionalParameters.java:250) - Optional Parameter Tag class org.jsmpp.bean.OptionalParameter$Sc_interface_version not found
> 2018-08-27 10:10:54,942 [SMS-Conn-Id-83] INFO  (SMPPSession.java:662) - Changing processor degree to 3
> 2018-08-27 10:10:54,942 [SMS-Conn-Id-83] ERROR (AbstractSessionContext.java:81) - Invalid runtime exception thrown when calling onStateChange for org.jsmpp.session.SMPPSession@491f0632
> java.lang.IllegalArgumentException
> 	at java.base/java.util.concurrent.ThreadPoolExecutor.setCorePoolSize(ThreadPoolExecutor.java:1560)
> 	at org.jsmpp.session.SMPPSession$BoundSessionStateListener.onStateChange(SMPPSession.java:663)
> 	at org.jsmpp.session.AbstractSessionContext.fireStateChanged(AbstractSessionContext.java:79)
> 	at org.jsmpp.session.SMPPSessionContext.changeState(SMPPSessionContext.java:61)
> 	at org.jsmpp.session.AbstractSessionContext.bound(AbstractSessionContext.java:51)
> 	at org.jsmpp.session.SMPPSession.connectAndBind(SMPPSession.java:236)
> 	at org.jsmpp.session.SMPPSession.connectAndBind(SMPPSession.java:201)
> 	at org.jsmpp.session.SMPPSession.<init>(SMPPSession.java:140)
> 	at cl.combase.ida.util.sms.AutoReconnectGateway.newSession(AutoReconnectGateway.java:117)
> 	at cl.combase.ida.util.sms.AutoReconnectGateway.<init>(AutoReconnectGateway.java:66)
> 	at cl.combase.ida.util.thread.ConnectorThread.run(ConnectorThread.java:91)
> 2018-08-27 10:10:54,944 [EnquireLinkSender: org.jsmpp.session.SMPPSession@491f0632] DEBUG (AbstractSession.java:457) - Starting EnquireLinkSender for session c1d3ad2f
```

It is very strange, because the processor degree is set to 3, but it throws the exception anyway.

Any idea?


Thanks,

**Comment by PascalSchumacher** [comment]
I guess this can be closed now that #92 was merged and release as part of jsmpp `2.3.7`?

**Comment by pmoerenhout** [comment]
Yes, it should have been resolved with release 2.3.7.
